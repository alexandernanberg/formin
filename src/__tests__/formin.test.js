import React, { StrictMode } from 'react'
import { render, fireEvent, act } from 'react-testing-library'
import 'jest-dom/extend-expect'
import { Formin } from '..'

jest.useFakeTimers()

function setup({ renderFn, formProps, ...props } = {}) {
  function defaultRenderFn({ getFormProps, getInputProps }) {
    return (
      <form data-testid="form" {...getFormProps(formProps)}>
        <input
          data-testid="input"
          required
          {...getInputProps({ name: 'name' })}
        />
        <input
          type="number"
          data-testid="number"
          {...getInputProps({ name: 'age' })}
        />
        <select data-testid="select" {...getInputProps({ name: 'pet' })}>
          <option value="dog">Dog</option>
          <option value="cat">Cat</option>
        </select>
        <input
          data-testid="checkbox"
          type="checkbox"
          {...getInputProps({ name: 'toys' })}
        />
        <button type="submit" data-testid="button" />
      </form>
    )
  }

  let renderArg
  const children = renderFn || defaultRenderFn
  const childrenSpy = jest.fn(controllerArg => {
    renderArg = controllerArg
    return children(controllerArg)
  })

  const utils = render(<Formin {...props}>{childrenSpy}</Formin>)

  return {
    ...utils,
    renderArg,
    childrenSpy,
    form: utils.queryByTestId('form'),
    input: utils.queryByTestId('input'),
    number: utils.queryByTestId('number'),
    checkbox: utils.queryByTestId('checkbox'),
    button: utils.queryByTestId('button'),
  }
}

test('basic snapshot', () => {
  const { container } = setup()
  expect(container.firstChild).toMatchSnapshot()
})

test('onChange called with changes', () => {
  const handleOnChange = jest.fn()
  const { input } = setup({
    onChange: handleOnChange,
  })

  fireEvent.change(input, { target: { value: 'Charlie' } })

  expect(handleOnChange).toHaveBeenCalledWith(
    expect.objectContaining({ name: 'Charlie' }),
  )
})

test('onSubmit called with stateAndHelpers', () => {
  const onSubmit = jest.fn(({ values }) => values)
  const { renderArg, form } = setup({
    onSubmit,
  })

  const { getInputProps, getFormProps, ...args } = renderArg

  fireEvent.submit(form)

  expect(onSubmit.mock.calls[0][0]).toEqual(args)
})

test.skip('onFocus sets touched', () => {
  const handleStateChange = jest.fn()
  const { input } = setup({
    onStateChange: handleStateChange,
  })

  fireEvent.focus(input)

  expect(handleStateChange).toHaveBeenCalledWith(
    expect.objectContaining({ touched: { name: true } }),
  )
})

test('can be controlled', () => {
  const onChange = jest.fn()
  const { input } = setup({ onChange, values: { name: 'Alex' } })

  expect(input.value).toEqual('Alex')

  fireEvent.change(input, { target: { value: 'Charlie' } })

  expect(onChange).toHaveBeenCalledWith({ name: 'Charlie' })
})

test('can reset', () => {
  const { input, renderArg } = setup({ defaultValues: { name: 'Alex' } })

  expect(input.value).toEqual('Alex')

  act(() => {
    renderArg.reset()
  })

  expect(input.value).toEqual('')
})

test('sets error onInvalid', () => {
  const { form, input } = setup()

  form.checkValidity()

  act(() => {
    jest.runAllTimers()
  })

  expect(input).toHaveAttribute('aria-invalid')
})

test('clear error onChange', () => {
  const { form, input } = setup()

  act(() => {
    form.checkValidity()
    jest.runAllTimers()
  })

  fireEvent.change(input, { target: { value: 'Charlie' } })

  expect(input).not.toHaveAttribute('aria-invalid')
})

test('set correct checkbox value', () => {
  const onChange = jest.fn()
  const { checkbox } = setup({
    onChange,
  })

  fireEvent.click(checkbox)

  expect(onChange).toHaveBeenCalledWith(expect.objectContaining({ toys: true }))
})

test('should work in StrictMode without warnings', () => {
  const spy = jest.spyOn(console, 'error').mockImplementation(() => {})

  render(
    <StrictMode>
      <Formin>
        {({ getFormProps, getInputProps }) => (
          <form {...getFormProps()}>
            <input type="text" {...getInputProps({ name: 'input' })} />
          </form>
        )}
      </Formin>
    </StrictMode>,
  )

  expect(spy).not.toHaveBeenCalled()
})

// TODO: Check focused element on error (should be first invalid)
