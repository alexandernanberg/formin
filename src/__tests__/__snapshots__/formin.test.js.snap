// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<form
  data-testid="form"
>
  <input
    data-testid="input"
    name="text"
    required=""
    value=""
  />
  <input
    data-testid="number"
    name="number"
    type="number"
    value=""
  />
  <select
    data-testid="select"
    name="select"
  >
    <option
      value="foo"
    >
      Foo
    </option>
    <option
      value="bar"
    >
      Bar
    </option>
  </select>
  <input
    data-testid="checkbox"
    name="checkbox"
    type="checkbox"
    value=""
  />
  <input
    data-testid="range"
    name="range"
    type="range"
    value=""
  />
  <input
    data-testid="custom"
    name="custom"
    type="text"
    value=""
  />
  <button
    data-testid="button"
    type="submit"
  />
</form>
`;
